### DISCUSSION TITLE
Batch outputs as one message

### DISCUSSION THREAD

**Discussion by Kilvoctu** [Kilvoctu/aiyabot#153] [Show and tell]
Heya, a few people have expressed that they don't like AIYA sending batch outputs as separate messages (one image per message). While I understand the amount of clutter it can add in practice, and I'm also against clutter, this is how I want to do things at the moment as I continue to develop the feature.  
I'd like to have more control over each image in the output, and also personally I just dislike Discord's new gallery view for multiple images (it makes weird grids and cuts off the image). If I'm able to address both of those points, then I may send batch outputs in one message again.

Anyway, in the meantime if you like AIYA sending batch outputs as one message, please clone this branch instead.  
**Please note that I won't be updating or supporting this branch.**  
https://github.com/Kilvoctu/aiyabot/tree/batch-alt  
It should look like before and additionally include any features/changes made up until now.   
The quirks are that you have a batch limit 10, and 🖋️ and 📋 are based on the first image.  
To continue generating in batches, you need to set it in [config](https://github.com/Kilvoctu/aiyabot/wiki/Configuration) (`batch_buttons`).

<details>
<summary>Example preview</summary>
<br>

![image](https://user-images.githubusercontent.com/2993060/226588256-172ad313-67d2-4805-a70c-7e13d3307a11.png)

</details>

**Comment by VictorZakharov** [comment]
Why not make it configurable instead? Like everything else about Aiya. Having more control is nice, but instead of seeing all pictures on a single (half) screen, I now need to scroll 2-3 screens up and down. This is a huge waste of mental resources.

**Reply by VictorZakharov** [reply]
@Kilvoctu Regarding buttons, what about showing buttons 1-4 (in case of batch/count size 4) ?
Then user clicks on 1,2,3,4 and Aiya uploads this image again with the usual controls.
Discord has icons for 1-9, I just checked, you can use those.

**Reply by Kilvoctu** [reply]
> Why not make it configurable instead? Like everything else about Aiya. Like everything else about Aiya. Having more control is nice, but instead of seeing all pictures on a single (half) screen, I now need to scroll 2-3 screens up and down. This is a huge waste of mental resources.

Of course I thought about this; however, the foundation of AIYA's code which handles Discord posting is changed. It isn't like recent commit https://github.com/Kilvoctu/aiyabot/commit/4b893bda736fc153add0bb9f60a9baccec4faece where an if statement and few lines of code was all that's need. I'd need to do a fairly major code refactor in that regard before this can be considered.

> Regarding buttons, what about showing buttons 1-4 (in case of batch/count size 4) ?
Then user clicks on 1,2,3,4 and Aiya uploads this image again with the usual controls.
Discord has icons for 1-9, I just checked, you can use those.

In the code, each message has a "Discord View" attached to it.
Discord Views are pre-defined with the "Discord Buttons". I can't dynamically add/remove buttons when initializing the View. I've already tried this for /info's buttons. That's why that command has ◀️▶️ buttons visible at all times even when not needed.  
Any batches from 2 to X would each need a customized View coded for it, which is just not practical. I don't think this is a viable way to go about it.

There are some ideas that have been circulating in my head on how to handle this issue, but I don't have time currently to try them out due to real life circumstances as of late... I created the alt branch so people can use it while I work things out.

Please remember I am a real person, a random non-developer nobody, working on AIYA alone in my free time.
I very much care about everyone's concerns, dedicating a large amount of my life into AIYA for the past 6 months addressing all feedback, without much in the way of thanks, contributions or support. It's fine, but it's not a pace or environment I can keep up with at times.  
Even in current circumstances, I still try to reply to all feedback and noting stuff in the [to-do projects](https://github.com/Kilvoctu/aiyabot/projects?query=is%3Aopen).

**Reply by VictorZakharov** [reply]
@Kilvoctu There is no need for a customized/modifiable view. Just show buttons 1-9, then when user clicks on any of them, it will copy/paste that image as if it was just generated and show the normal UI for it. This way there is minimal work. So just two different but fixed UIs for 1 image and 2 to X. If you need it to perform an operation, do a noop like upscale x1.0 with the source url.

In software, when something needs to be done, you do not necessarily drop everything and do it. There is priority and planning, then an iterative approach where features are delivered in small chunks and each can be verified on its own. As a software architect, I can only suggest a direction.

Compared to other open source projects, Aiya has improved a lot in the past few months, I appreciate your effort in this.

**Comment by sebaxakerhtc** [comment]
> Anyway, in the meantime if you like AIYA sending batch outputs as one message, please clone this branch instead. **Please note that I won't be updating or supporting this branch.** https://github.com/Kilvoctu/aiyabot/tree/batch-alt 

Upscaler is not work on this branch for me.
It do upscale, save it, but don't send
```
Saved image: outputs/1682240646-x2.0-.png
Exception in thread Thread-50 (post_dream):
Traceback (most recent call last):
  File "C:\Users\sebaxakerhtc\AppData\Local\Programs\Python\Python310\lib\threading.py", line 1009, in _bootstrap_inner
    self.run()
  File "C:\Users\sebaxakerhtc\AppData\Local\Programs\Python\Python310\lib\threading.py", line 946, in run
    self._target(*self._args, **self._kwargs)
  File "C:\aiyabot\core\upscalecog.py", line 220, in post_dream
    self, queuehandler.PostObject(
TypeError: PostObject.__init__() missing 1 required positional argument: 'files'
```
What to do to make it working?

**Reply by Kilvoctu** [reply]
Heya, in upscalecog.py, on line 221, replace the text with this:  
`self, queue_object.ctx, content=f'<@{queue_object.ctx.author.id}>, {message}', files='', file=file, embed='', view=queue_object.view))`  

Or pull that branch again, I pushed an update to fix it.

**Reply by sebaxakerhtc** [reply]
Yeah, edited and it work again! Thank you!

One more question - where i can change the Upscale default Strength from 0.5 to 0.35?

Upd:
I'm not sure is the right way, but i changed this value and work for me
`upscaler_2_strength: Optional[str] = '0.35',` on line 95 of upscalecog.py 
and 
`upscaler_2, upscaler_2_strength = "None", '0.35'` on line 245 of ctxmenuhandler.py

This is strange that the upscaler 2 value is the same for upscaler 1 even we don't use upscaler 2
I think it's better to add `Strength` to options in /upscale command like we have it in /draw

And Great thank you for this project!

**Reply by Kilvoctu** [reply]
Yeah that would be the correct value to change.  

----

For upscaler 2, it's not supposed to even send  upscaler 2 in the payload if it's not set, due to this section in upscalecog.py, starting at line 183:  
```
if queue_object.upscaler_2 is not None:
    up2_payload = {
        "upscaler_2": queue_object.upscaler_2,
         "extras_upscaler_2_visibility": queue_object.upscaler_2_strength
    }
    payload.update(up2_payload)
```

However, I'm seeing the issue in my code that's causing an issue. `if queue_object.upscaler_2 is not None` is looking for if any value is set for `upscaler_2`. The issue is that by default, I have `upscaler_2` set to the string `"None"`.  

You can fix this by changing these lines to:
line 128: change to `if upscaler_2 != "None":`
line 183: change to `if queue_object.upscaler_2 != "None":`

----

`upscaler_2_strength`, in the Web UI, actually is called `upscaler_2_visibility`. I just named it strength to minimize the amount of terms that end user would need to keep in mind.  

The way the upscale works is that `upscaler_1` is always applied at 100%.  
When `upscaler_2` is added, now the upscaling is split between two models. So with my default value, it would be 50% upscaler 1, 50% upscaler 2.   
`upscaler_2_strength` is how the end user adjusts this percentage. If user sets, for example, `upscaler_2_strength` to 1, then the split would be 0% upscaler 1, 100% upscaler 2.   

Probably I could communicate this behavior better in-app 🤔

----

I'm glad to hear you're enjoying the aiyabot. I really wish I could work on her more, but it's hard to find the time and motivation recently as real life is hitting me more. But of course, I still read and address all feedback the best I can.

**Reply by sebaxakerhtc** [reply]
> You can fix this by changing these lines to: line 128: change to `if upscaler_2 != "None":` line 183: change to `if queue_object.upscaler_2 != "None":`

Thank you for your help!
Edited code like you mentioned, but Upscaler 2 sill isn't used.
And, of course, the right way is to have Upscaler 1 strength and Upscaler 2 visibility.

Cheers!

**Reply by Kilvoctu** [reply]
I tested the branch, and Upscaler 2 seems to be used 🤔. Not sure what you may be doing differently.  
In this quick image, I put in my image and selected `ESRGAN_4x` as my Upscaler 2. AIYA tells me my options in initial reply.  
I took the output and put it in Web UI PNG Info, and it confirms that it did process said options as requested.     
![image](https://user-images.githubusercontent.com/2993060/234105503-552860c0-1ef5-4820-ba27-f2f5f100b354.png)

Upscaler 1 strength doesn't exist. That is, Web UI extras upscale doesn't have any parameter for it. Upscaler 1 is at the maximum strength, unless upscaler 2 is used.

**Reply by sebaxakerhtc** [reply]
I looked into terminal window and 2nd upscaler is not persist

**Reply by Kilvoctu** [reply]
What does the terminal window say?

**Reply by sebaxakerhtc** [reply]
> What does the terminal window say?

Say nothing 😁
There's something wrong on my side
Deleted folder and cloned repo again.
`git clone -b batch-alt https://github.com/Kilvoctu/aiyabot.git`
The result:
![1](https://user-images.githubusercontent.com/32651506/234181421-466b9fec-f947-4079-b406-98ea69e19609.png)
![2](https://user-images.githubusercontent.com/32651506/234181446-13ea442b-6017-42de-b6b4-09a5d68c213c.png)

**Reply by Kilvoctu** [reply]
Discord recently started removing all the png info from images uploaded to their app. If you save the original image to disk, it should have the information. If AIYA is showing in the confirmation that Upscaler 2 is used, it should be working though 🤔

**Reply by sebaxakerhtc** [reply]
> Discord recently started removing all the png info from images uploaded to their app. If you save the original image to disk, it should have the information. If AIYA is showing in the confirmation that Upscaler 2 is used, it should be working though 🤔

Yes, it works.
